src/EditorFeatures/Core/Intents/RenameIntentProvider.cs

```diff
@@ -0,0 +1,55 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Editor;
+using Microsoft.CodeAnalysis.Features.Intents;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Rename;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.EditorFeatures.Intents;
+
+[IntentProvider(WellKnownIntents.Rename, LanguageNames.CSharp), Shared]
+internal class RenameIntentProvider : IIntentProvider
+{
+    private record RenameIntentData(string NewName);
+
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public RenameIntentProvider()
+    {
+    }
+
+    public async Task<ImmutableArray<IntentProcessorResult>> ComputeIntentAsync(Document priorDocument, TextSpan priorSelection, Document currentDocument, IntentDataProvider intentDataProvider, CancellationToken cancellationToken)
+    {
+        var renameIntentData = intentDataProvider.GetIntentData<RenameIntentData>();
+        Contract.ThrowIfNull(renameIntentData);
```

Comment:
is this like a json deserialization thing?

Reply:
yup, and the string they pass us is nullable in the API so we have to account for it